src/ray/core_worker/transport/direct_task_transport.h

```diff
@@ -0,0 +1,116 @@
+#ifndef RAY_CORE_WORKER_DIRECT_TASK_H
+#define RAY_CORE_WORKER_DIRECT_TASK_H
+
+#include "absl/base/thread_annotations.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/common/id.h"
+#include "ray/common/ray_object.h"
+#include "ray/core_worker/context.h"
+#include "ray/core_worker/store_provider/memory_store_provider.h"
+#include "ray/core_worker/transport/direct_actor_transport.h"
+#include "ray/raylet/raylet_client.h"
+#include "ray/rpc/worker/core_worker_client.h"
+
+namespace ray {
+
+struct TaskState {
+  /// The task to be run.
+  const TaskSpecification task;
+  /// The remaining dependencies to resolve for this task.
+  absl::flat_hash_set<ObjectID> local_dependencies;
+};
+
+// This class is thread-safe.
+class LocalDependencyResolver {
+ public:
+  LocalDependencyResolver(RayletClient &raylet_client,
+                          CoreWorkerMemoryStoreProvider &store_provider)
+      : raylet_client_(raylet_client), in_memory_store_(store_provider) {}
+
+  /// Resolve all local and remote dependencies for the task, calling the specified
+  /// callback when done. Direct call ids in the task specification will be resolved
+  /// to concrete values and inlined.
+  //
+  /// Note: This method **will mutate** the given TaskSpecification.
+  ///
+  /// Postcondition: all direct call ids in arguments are converted to values.
+  void ResolveDependencies(const TaskSpecification &task,
+                           std::function<void()> on_complete);
+
+ private:
+  // Reference to the shared raylet client for waiting for deps.
+  RayletClient &raylet_client_;
+
+  /// The store provider.
+  CoreWorkerMemoryStoreProvider &in_memory_store_;
+
+  /// Protects against concurrent access to internal state.
+  absl::Mutex mu_;
+};
+
+typedef std::pair<std::string, int> WorkerAddress;
+
+// This class is thread-safe.
+class CoreWorkerDirectTaskSubmitter {
+ public:
+  CoreWorkerDirectTaskSubmitter(
+      RayletClient &raylet_client, rpc::ClientCallManager &client_call_manager,
+      std::unique_ptr<CoreWorkerMemoryStoreProvider> store_provider)
+      : raylet_client_(raylet_client),
+        client_call_manager_(client_call_manager),
+        in_memory_store_(std::move(store_provider)),
+        resolver_(raylet_client, *in_memory_store_) {}
+
+  /// Schedule a task for direct submission to a worker.
+  Status SubmitTask(const TaskSpecification &task_spec);
+
+  /// Callback for when the raylet grants us a worker lease.
```

Comment:
Add a comment explaining how/when to return the lease.

Reply:
Done.